File: Python/Stacks & Queues/34_maximum_of_all_subarray.py
"""
Idea:
Implementing sliding window technique, using Double ended Queue to keep track of the max element positions.
Inserting at the end, such that inserting element is greater than the end.
"""

# User function Template for python3
from collections import deque


class Solution:

    # Function to find maximum of each subarray of size k.
    def max_of_subarrays(self, arr, n, k):

        ans = []
        dq = deque()

        # Process the first window separately
        for i in range(k):
            while dq and arr[i] >= arr[dq[-1]]:
                dq.pop()
            dq.append(i)

        # Process the remaining windows
        for i in range(k, n):
            ans.append(arr[dq[0]])

            # Remove elements that are out of the current window
            while dq and dq[0] <= i - k:
                dq.popleft()

            # Remove smaller elements as they will not be the maximum
            while dq and arr[i] >= arr[dq[-1]]:
                dq.pop()

            dq.append(i)

        # Add the maximum of the last window
        ans.append(arr[dq[0]])

        return ans


# {
# Driver Code Starts
# Initial Template for Python 3

import atexit
import io
import sys
from collections import deque


_INPUT_LINES = sys.stdin.read().splitlines()
input = iter(_INPUT_LINES).__next__
_OUTPUT_BUFFER = io.StringIO()
sys.stdout = _OUTPUT_BUFFER


@atexit.register
def write():
    sys.__stdout__.write(_OUTPUT_BUFFER.getvalue())


if __name__ == '__main__':
    test_cases = int(input())
    for cases in range(test_cases):
        n, k = map(int, input().strip().split())
        arr = list(map(int, input().strip().split()))
        ob = Solution()
        res = ob.max_of_subarrays(arr, n, k)
        for i in range(len(res)):
            print(res[i], end=" ")
        print()
# } Driver Code Ends

File: C++/Stacks & Queues/34_maximum_of_all_subarray.cpp
/*
Idea:
Implementing sliding window technique, using Double ended Queue to keep track of the max element positions.
Inserting at the end, such that inserting element is greater than the end.
*/

//{ Driver Code Starts
//Initial template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
//User function template for C++

class Solution
{
public:
    //Function to find maximum of each subarray of size k.
    vector <int> max_of_subarrays(int* arr, int n, int k)
    {
        vector<int> ans;
        deque<int> dq;

        // Process the first window separately
        for (int i = 0; i < k; i++)
        {
            while (!dq.empty() && arr[i] >= arr[dq.back()])
                dq.pop_back();
            dq.push_back(i);
        }

        // Process the remaining windows
        for (int i = k; i < n; i++)
        {
            ans.push_back(arr[dq.front()]);

            // Remove elements that are out of the current window
            while (!dq.empty() && dq.front() <= i - k)
                dq.pop_front();

            // Remove smaller elements as they will not be the maximum
            while (!dq.empty() && arr[i] >= arr[dq.back()])
                dq.pop_back();

            dq.push_back(i);
        }

        // Add the maximum of the last window
        ans.push_back(arr[dq.front()]);

        return ans;
    }
};

//{ Driver Code Starts.

int main() {

    int t;
    cin >> t;

    while (t--) {

        int n, k;
        cin >> n >> k;

        int arr[n];
        for (int i = 0;i < n;i++)
            cin >> arr[i];
        Solution ob;
        vector <int> res = ob.max_of_subarrays(arr, n, k);
        for (int i = 0; i < res.size(); i++)
            cout << res[i] << " ";
        cout << endl;

    }

    return 0;
}
// } Driver Code Ends